xarray/core/missing.py

```diff
@@ -669,180 +684,172 @@ def interp(var, indexes_coords, method: InterpOptions, **kwargs):
-    x, new_x = _floatize_x(x, new_x)
-
-    if len(x) == 1:
-        return _interp1d(var, x, new_x, func, kwargs)
-
-    # move the interpolation axes to the start position
-    var = var.transpose(range(-len(x), var.ndim - len(x)))
-    # stack new_x to 1 vector, with reshape
-    xi = np.stack([x1.values.ravel() for x1 in new_x], axis=-1)
-    rslt = func(x, var, xi, **kwargs)
-    # move back the interpolation axes to the last position
-    rslt = rslt.transpose(range(-rslt.ndim + 1, 1))
-    return reshape(rslt, rslt.shape[:-1] + new_x[0].shape)
-
-
-def _chunked_aware_interpnd(var, *coords, interp_func, interp_kwargs, localize=True):
-    """Wrapper for `_interpnd` through `blockwise` for chunked arrays.
-
+def _interpnd(
+    data: np.ndarray,
+    *coords: np.ndarray,
+    interp_func: Interpolator | InterpCallable,
+    interp_kwargs,
+    result_coord_core_dims: list[tuple[Hashable, ...]],
+) -> np.ndarray:
+    """
+    Core nD array interpolation routine.
     The first half arrays in `coords` are original coordinates,
-    the other half are destination coordinates
+    the other half are destination coordinates.
     """
     n_x = len(coords) // 2
-    nconst = len(var.shape) - n_x
+    ndim = data.ndim
+    nconst = ndim - n_x
 
-    # _interpnd expect coords to be Variables
+    # Convert everything to Variables, since that makes applying
+    # `_localize` and `_floatize_x` much easier
     x = [Variable([f"dim_{nconst + dim}"], _x) for dim, _x in enumerate(coords[:n_x])]
-    new_x = [
-        Variable([f"dim_{len(var.shape) + dim}" for dim in range(len(_x.shape))], _x)
-        for _x in coords[n_x:]
-    ]
-
-    if localize:
-        # _localize expect var to be a Variable
-        var = Variable([f"dim_{dim}" for dim in range(len(var.shape))], var)
+    new_x = list(
+        broadcast_variables(
+            *(
+                Variable(dims, _x)
+                for dims, _x in zip(result_coord_core_dims, coords[n_x:], strict=True)
+            )
+        )
+    )
+    var = Variable([f"dim_{dim}" for dim in range(ndim)], data)
 
+    if interp_kwargs.get("method") in ["linear", "nearest"]:
         indexes_coords = {
             _x.dims[0]: (_x, _new_x) for _x, _new_x in zip(x, new_x, strict=True)
         }
-
         # simple speed up for the local interpolation
         var, indexes_coords = _localize(var, indexes_coords)
-        x, new_x = zip(*[indexes_coords[d] for d in indexes_coords], strict=True)
+        x, new_x = tuple(
+            list(_)
+            for _ in zip(*(indexes_coords[d] for d in indexes_coords), strict=True)
+        )
 
-        # put var back as a ndarray
-        var = var.data
+    x_list, new_x_list = _floatize_x(x, new_x)
 
-    return _interpnd(var, x, new_x, interp_func, interp_kwargs)
+    if len(x) == 1:
+        # TODO: narrow interp_func to interpolator here
+        return _interp1d(var, x_list, new_x_list, interp_func, interp_kwargs)  # type: ignore[arg-type]
```

Comment:
Mypy is correct to error here right? 
_interp1d calls interp_func(...)(....) and that should crash with a InterpCallable?
Is there a pytest with `interp_func: InterpCallable`? 
Is InterpCallable necessary? Would be nice to just remove it...

Reply:
it depends on whether we end up using `get_interpolator` or `get_interpolator_nd`. I'm sure there's a test but can't remember which off the top of my head.